test/cli/cli_test.cpp

```diff
@@ -45,9 +47,32 @@ struct cli_test : public ::testing::Test
 
     // Generate the full path of a test file that is provided in the data directory.
     static
-    std::string data(std::string const & filename)
+    std::filesystem::path data(std::string const & filename)
     {
-        return std::string{DATADIR}.append(filename);
+        return std::filesystem::path{std::string{DATADIR}}.concat(filename);
+    }
+
+    // Holds the original work directory.
+    std::filesystem::path original_wd{};
+
+    // Creates a work directory for the test.
+    void SetUp() override
+    {
+        ::testing::TestInfo const * const info = ::testing::UnitTest::GetInstance()->current_test_info();
+        std::filesystem::path const test_dir{std::string{OUTPUTDIR} +
+                                             std::string{info->test_case_name()} +
+                                             std::string{"."} +
+                                             std::string{info->name()}};
+
+        ASSERT_NO_THROW(std::filesystem::remove_all(test_dir));         // delete the directory if it exists
+        ASSERT_NO_THROW(std::filesystem::create_directories(test_dir)); // create new empty directory
+        ASSERT_NO_THROW(original_wd = std::filesystem::current_path()); // store original wd
+        ASSERT_NO_THROW(std::filesystem::current_path(test_dir));       // change wd
```

Comment:
Let's compare:

With ASSERT_NO_THROW it looks like this:
```
app-template/test/cli/cli_test.cpp:70: Failure
Expected: std::filesystem::current_path(test_dir) doesn't throw an exception.
  Actual: it throws std::exception-derived exception with description: "filesystem error: cannot set current path: No such file or directory".
```

without: 
```
unknown file: Failure
C++ exception with description "filesystem error: cannot set current path: No such file or directory" thrown in SetUp().
```

I like in the second version that it reports the SetUp() function, but we do not see the file and line where the failure occurs. The ASSERT_NO_THROW command can be extended with further information like
```cpp
ASSERT_NO_THROW() << "Failed to set " << test_dir << " as work directory."
```
which prints an extra line so the user can see which directory causes the problem. Therefore, I suggest to use the macro with additional information.

Then we see:
```
app-template/test/cli/cli_test.cpp:70: Failure
Expected: std::filesystem::current_path(test_dir) doesn't throw an exception.
  Actual: it throws std::exception-derived exception with description: "filesystem error: cannot set current path: No such file or directory".
Failed to set "build/app-template/test/output/cli_test.no_options" as work directory.
```

Reply:
Yep, I get your point. I have chosen assert over expect, because the test should not execute further if there is no work directory (that's the difference according to the gtest doc). 

I do not want to overcomplicate things, so I will just use the commands without any *_NO_THROW as you initially suggested. 

Last question @marehr: Can I put the commands in a try-catch block like the following or better not?
```cpp
try 
{
    4x std::filesystem commands without assert...
} 
catch (std::exception e)
{
    FAIL() << "Creating test directory " << test_dir << " failed: " + e.message();
}
```

Reply:
Sorry for the late reply. You can do that :)